.github/workflows/ci.yml

```diff
@@ -3,61 +3,53 @@ name: Continuous Integration
 on: [push, pull_request]
 
 jobs:
-  lint-and-setup:
+  setup-and-lint:
     name: Setup and Lint Rust Code
     runs-on: ubuntu-latest
     steps:
       - name: Check out code
-        uses: actions/checkout@v4
+        uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
         with:
           submodules: recursive
-      - name: Install dependencies and common setup
+      - name: Cache Project
+        uses: Swatinem/rust-cache@23bce251a8cd2ffc3c1075eaa2367cf899916d84  # v2.7.3
+      - name: Install toolchain
         uses: ./.github/actions/setup-env
+        with:
+          repo-token: ${{ secrets.GITHUB_TOKEN }}
+
       - name: Lint Rust code with rustfmt and clippy
         run: |
           cargo +nightly fmt
-          cargo +nightly clippy
-
-  build:
-    name: Build Rust Project
-    runs-on: ubuntu-latest
-    needs: lint-and-setup
-    steps:
-      - name: Check out code
-        uses: actions/checkout@v4
-        with:
-          submodules: recursive
-      - name: Install dependencies and common setup
-        uses: ./.github/actions/setup-env
-      - name: Build Rust project
-        run: cargo build
+          cargo clippy
 
-  test:
-    name: Run spiffe and spire-api Integration Tests
+  build-and-test:
+    name: Build and Test Rust Project
     runs-on: ubuntu-latest
+    needs: setup-and-lint
     env:
       SPIFFE_ENDPOINT_SOCKET: unix:/tmp/spire-agent/public/api.sock
       SPIRE_ADMIN_ENDPOINT_SOCKET: unix:/tmp/spire-agent/admin/api.sock
-    needs: build
     steps:
       - name: Check out code
-        uses: actions/checkout@v4
+        uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
```

Comment:
I assume one can't write `uses: actions/checkout@v4.1.1`?

Reply:
Indeed, you can use `actions/checkout@v4.1.1`; however, my approach prioritizes pinning the version using its commit SHA for enhanced security and stability. This method ensures that the exact version of the action is always used.